### PR TITLE
GitHub Action should warn users when the CSP API token is about to ex…

### PR DESCRIPTION
**Pull request by gdelgadot**
GitHub Action logs a warning about token expiration with a human-readable date.

Signed-off-by: Gloria Delgado <gloriad@vmware.com>

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -545,6 +546,30 @@ export async function getToken(input: CspInput): Promise<string> {
...
 
+export async function checkTokenExpiration(input: CspInput): Promise<string> {
+  const response = await cspClient.post("https://console.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/details", {
+    headers: {
+      "Content-Type": "application/json",
+      tokenValue: "$CSP_API_TOKEN",
+    },
+  })
+
+  cachedCspToken = {
```
Comment (mpermar) [inline src/main.ts:557]: I'm not sure you can use what you get back from that request as a token for talking to our APIs. Better not to cache it here.

File: src/main.ts [hunk 2]
```diff
@@ -545,6 +546,30 @@ export async function getToken(input: CspInput): Promise<string> {
...
+    access_token: response.data.access_token,
+    timestamp: Date.now() + input.timeout,
+  }
+
+  if (response.data.details) {
```
Comment (mpermar) [inline src/main.ts:562]: This logic looks wrong. It should be placed after the check that you have below.

File: src/main.ts [hunk 3]
```diff
@@ -545,6 +546,30 @@ export async function getToken(input: CspInput): Promise<string> {
...
+    return response.data.expiresAt
+  }
+
+  if (response.data.expiresAt < `${moment().add(1, "month").calendar()}`) {
+    core.warning(`CSP API token will expire in ${moment().add(1, "month").calendar()}.`)
+  }
+
+  return response.data.details
```
Comment (mpermar) [inline src/main.ts:570]: You need to be consistent in the return value. Above, you're returning `expireAt` and here you are returning `details`.

File: src/main.ts [hunk 4]
```diff
@@ -545,6 +546,30 @@ export async function getToken(input: CspInput): Promise<string> {
...
+    access_token: response.data.access_token,
+    timestamp: Date.now() + input.timeout,
+  }
+
+  if (response.data.details) {
+    return response.data.expiresAt
+  }
+
+  if (response.data.expiresAt < `${moment().add(1, "month").calendar()}`) {
+    core.warning(`CSP API token will expire in ${moment().add(1, "month").calendar()}.`)
```
Comment (mpermar) [inline src/main.ts:567]: I think we should show the actual number of days to expiration. I recall moment has some diff methods and also the way to show a diff in human-friendly language, i.e. 3d and 12 hours.

**Review by mpermar — CHANGES_REQUESTED** [review]

File: dist/licenses.txt [hunk 5]
```diff
@@ -356,6 +356,32 @@ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
 IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
 
+moment
```
Comment (recena) [inline dist/licenses.txt:359]: @gdelgadot This is an unrelated change.

File: src/main.ts [hunk 6]
```diff
@@ -545,6 +546,28 @@ export async function getToken(input: CspInput): Promise<string> {
   }
 }
 
+export async function checkTokenExpiration(): Promise<string> {
+  const response = await cspClient.post("https://console.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/details", {
```
Comment (recena) [inline src/main.ts:550]: Could we define a constant to centralize the configuration?

File: dist/licenses.txt (hunk 5 above)
Comment (gdelgadot) [reply dist/licenses.txt:359]: yes, actually I didn't notice it

Comment (gdelgadot) [reply dist/licenses.txt:359]: but it has been added because of the installation of `moment.js`

File: src/main.ts (hunk 6 above)
Comment (gdelgadot) [reply src/main.ts:550]: do you mean a constant like the default base folder or pipeline in `constants.ts`?

File: dist/licenses.txt (hunk 5 above)
Comment (mpermar) [reply dist/licenses.txt:359]: This is part of the typescript packaging.

File: src/main.ts (hunk 6 above)
Comment (mpermar) [reply src/main.ts:550]: Actually good catch. You need to remove the domain name and keep just the URL path, [like here](https://github.com/vmware-labs/vmware-image-builder-action/blob/main/src/main.ts#L527).

Then, a good improvement is what @recena suggests, moving the paths (e.g. `/csp/gateway/am/api/auth/api-tokens/details`) to constants, e.g. (`constants.TOKEN_DETAILS_PATH`)

Comment (gdelgadot) [reply src/main.ts]: It should be done 🙂

File: package.json [hunk 7]
```diff
@@ -45,10 +46,10 @@
     "glob": "^7.2.0",
     "jest": "^27.2.5",
     "jest-circus": "^27.4.5",
+    "minimist": ">=1.2.6",
```
Comment (mpermar) [inline package.json:49]:
```suggestion
    "minimist": "^1.2.6",
```

File: src/constants.ts [hunk 8]
```diff
@@ -83,3 +83,8 @@ export enum RetriableHttpStatus {
  * Prefix for environment variables that will be used for template substitution in pipelines.
  */
 export const ENV_VAR_TEMPLATE_PREFIX = "VIB_ENV_"
+
+/**
+ * CSP endpoint to get API token details
+ */
+export const TOKEN_DETAILS_PATH = "/csp/gateway/am/api/auth/api-tokens/details"
```
Comment (mpermar) [inline src/constants.ts:90]: Now that you have done this, can you do it also for the other paths with use with `cspClient`?

**Comment by gdelgadot** [comment]
I'm having several issues to update this old branch, so I will close the PR and open a new one
